bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,402 @@
+    implements SupportsNamespaces, Configurable {
+
+  // User provided properties.
+  public static final String PROPERTIES_KEY_GCP_PROJECT = "gcp-project";
+  public static final String PROPERTIES_KEY_GCP_LOCATION = "gcp-location";
+  public static final String PROPERTIES_KEY_FILTER_UNSUPPORTED_TABLES = "filter-unsupported-tables";
+  public static final String PROPERTIES_KEY_TESTING_ENABLED = "testing-enabled";
+
+  public static final String HIVE_METASTORE_WAREHOUSE_DIR = "hive.metastore.warehouse.dir";
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryMetastoreCatalog.class);
+
+  private static final String DEFAULT_GCP_LOCATION = "us";
+
+  private String catalogPluginName;
+  private Map<String, String> catalogProperties;
+  private FileIO fileIO;
+  private Configuration conf;
+  private String projectId;
+  private String location;
+  private BigQueryMetastoreClient client;
+  private boolean filterUnsupportedTables;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigQueryMetastoreCatalog() {}
+
+  @Override
+  public void initialize(String inputName, Map<String, String> properties) {
+    if (!properties.containsKey(PROPERTIES_KEY_GCP_PROJECT)) {
+      throw new ValidationException("GCP project must be specified");
+    }
+    projectId = properties.get(PROPERTIES_KEY_GCP_PROJECT);
+    location = properties.getOrDefault(PROPERTIES_KEY_GCP_LOCATION, DEFAULT_GCP_LOCATION);
+    boolean testingEnabled =
+        Boolean.parseBoolean(properties.getOrDefault(PROPERTIES_KEY_TESTING_ENABLED, "false"));
+
+    BigQueryOptions options =
+        BigQueryOptions.newBuilder()
+            .setProjectId(projectId)
+            .setLocation(location)
+            .setRetrySettings(ServiceOptions.getDefaultRetrySettings())
+            .build();
+
+    try {
+      if (testingEnabled) {
+        client = new FakeBigQueryMetastoreClient(options);
+
+      } else {
+        client = new BigQueryMetastoreClientImpl(options);
+      }
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigQuery client failed");
+    } catch (GeneralSecurityException e) {
+      throw new ValidationException(e, "Creating BigQuery client failed due to a security issue");
+    }
+    initialize(inputName, properties, projectId, location, client);
+  }
+
+  @VisibleForTesting
+  void initialize(
+      String inputName,
+      Map<String, String> properties,
+      String initialProjectId,
+      String initialLocation,
+      BigQueryMetastoreClient bigQueryMetaStoreClient) {
+    this.catalogPluginName = inputName;
+    this.catalogProperties = ImmutableMap.copyOf(properties);
+    this.projectId = initialProjectId;
+    this.location = initialLocation;
+    this.client =
+        Preconditions.checkNotNull(bigQueryMetaStoreClient, "BigQuery client can not be null");
```

Comment:
IMO it's better to use `Preconditions.checkArgument(null != ..., "Invalid BigQuery client: null")` because it indicates that a wrong argument was passed in a clearer way than a NPE